grammarinator/generate.py

```diff
@@ -126,12 +121,16 @@ def create_new_test(self, index, lock):
             logger.warning('Test generation failed.', exc_info=e)
             return self.create_new_test(index, lock)
 
-        test_fn = self.out_format % index if self.out_format else None
+        test_fn = self.out_format % index if '%d' in self.out_format else self.out_format
         tree.root = Generator.transform(tree.root, self.transformers)
 
         tree_fn = None
         if self.population and self.keep_trees:
-            tree_fn = join(self.population.directory, basename(test_fn) + Tree.extension)
+            tree_basename = basename(test_fn)
+            if '%d' not in tree_basename:
```

Comment:
By the time execution gets here, `test_fn` contains no `%d` (under normal circumstances), so `tree_basename` will contain none either. See line 124: if `out_format` contains `%d`, then it will be substituted by `index`, so `test_fn` will be `%d`-free. So, I guess that `tree_basename` will have to be constructed from `basename(self.out_format)`.